### PR TITLE
Add instructions for creating custom system image

### PR DESCRIPTION
**Pull request by jkozdon**
CLIMA startup times are long, which can make iterative development and
debugging difficult / painful. This adds a process for creating a custom
system image which drastically reduces initialization time.

|                      command                      |  no image |  with image |
|---------------------------------------------------|-----------|-------------|
| experiments/AtmosLES/dycoms.jl --help             |   31.44s  |       2.5s  |
| experiments/AtmosLES/dycoms.jl                    |  244.03s  |    138.64s  |
| test/Numerics/DGmethods/Euler/isentropicvortex.jl |  229.62s  |    144.19s  |

I have

- [x] Written and run all necessary tests with CLIMA by including `tests/runtests.jl`
- [x] Followed all necessary [style guidelines](https://climate-machine.github.io/CLIMA/latest/CodingConventions.html) and run `julia .dev/format.jl`
- [x] Updated the documentation to reflect changes from this PR.

# For review by CLIMA developers

- [x] There are no open pull requests for this already
- [x] CLIMA developers with relevant expertise have been assigned to review this submission
- [x] The code conforms to the [style guidelines](https://climate-machine.github.io/CLIMA/latest/CodingConventions.html) and has consistent naming conventions. `julia .dev/format.jl` has been run in a separate commit.
- [x] This code does what it is technically intended to do (all numerics make sense physically and/or computationally)

### CODE REVIEW

File: .dev/systemimage/climate_machine_image.jl [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+# specified by the argument (relative to the callers path)
+sysimage_path =
+    isempty(ARGS) ? joinpath(@__DIR__, "ClimateMachine.so") : ARGS[1]
+
+using Pkg
+Pkg.add("PackageCompiler")
+using PackageCompiler
+Pkg.activate(joinpath(@__DIR__, "../../"))
+Pkg.instantiate()
+pkgs = Pkg.installed()
```
Comment (jakebolewski) [inline .dev/systemimage/climate_machine_image.jl:17]: Should this be constrained to just the whitelisted CliMA project dependencies? (ex. a user installs additional packages into their project environment could lead to divergent system images maybe causing confusion / debugging issues)

Comment (jkozdon) [reply .dev/systemimage/climate_machine_image.jl:17]: The installed only includes packages from the project.

As seen below after the `Pkg.activate` both PackageCompiler and Revise are not listed by `Pkg.installed`

```
❯ julia -e "using Pkg; display(Pkg.installed()); println(); println();Pkg.activate(@__DIR__); println(); display(Pkg.installed());"
Dict{String,Union{Nothing, VersionNumber}} with 2 entries:
  "PackageCompiler" => v"1.1.1"
  "Revise" => v"2.6.4"

Activating environment at `~/codes/CLIMA/Project.toml`

Dict{String,Union{Nothing, VersionNumber}} with 34 entries:
  "Pkg" => nothing
  "Statistics" => nothing
  "ForwardDiff" => v"0.10.10"
  "Coverage" => v"1.0.0"
  "Distributions" => v"0.23.2"
  "MPI" => v"0.13.0"
  "CLIMAParameters" => v"0.1.3"
  "Random" => nothing
  "GaussQuadrature" => v"0.5.4"
  "CUDAdrv" => v"6.0.0"
  "OrderedCollections" => v"1.1.0"
  "Literate" => v"2.3.1"
  "DelimitedFiles" => nothing
  "LinearAlgebra" => nothing
  "DoubleFloats" => v"1.1.8"
  "GPUifyLoops" => v"0.2.9"
  "Combinatorics" => v"1.0.0"
  "CUDAapi" => v"3.1.0"
  "LazyArrays" => v"0.16.5"
  "CuArrays" => v"1.7.3"
  "DocStringExtensions" => v"0.8.1"
  "WriteVTK" => v"1.4.1"
  "CUDAnative" => v"2.10.2"
  "FileIO" => v"1.2.4"
  "StaticArrays" => v"0.12.1"
  "JLD2" => v"0.1.13"
  "NCDatasets" => v"0.10.2"
  "Dates" => nothing
  "SpecialFunctions" => v"0.10.0"
  "Logging" => nothing
  "Printf" => nothing
  "KernelAbstractions" => v"0.2.3"
  "RootSolvers" => v"0.2.0"
  "ArgParse" => v"1.1.0"
```

**Comment by simonbyrne** [comment]
As a future reference point, we should also look at which methods we invalidate: see https://github.com/JuliaLang/julia/pull/35714

**Comment by jakebolewski** [comment]
Was able to reproduce, -> ~2s startup. It doesn't seem like the generated PackageCompiler system image checks if the Julia minor run time versions are the same though in the so file (leads to segfault).

**Comment by jkozdon** [comment]
> Was able to reproduce, -> ~2s startup. It doesn't seem like the generated PackageCompiler system image checks if the Julia minor run time versions are the same though in the generated so file (leads to segfault).

Good to know. Once CLIMA is running on 1.4 we'll need to change this anyway since [`installed` is begin deprecated](https://github.com/JuliaLang/Pkg.jl/blob/127412cd23224a4f8fcdb049d80c4c16486a1e93/src/Pkg.jl#L531-L541) in favor of [`dependencies`](https://julialang.github.io/Pkg.jl/dev/api/#Pkg.dependencies)

**Comment by jkozdon** [comment]
The other thing I do not like about this is that using `isentropic_vortex` is a bit heavy of an initialization file (don't need to run that long). Open to other suggestions.

**Comment by kpamnany** [comment]
IMO this should be in v0.1.

File: .dev/systemimage/climate_machine_image.jl [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+delete!(pkgs, "Pkg")
+
+PackageCompiler.create_sysimage(
+    [Symbol(s) for s in keys(pkgs)];
+    sysimage_path = sysimage_path,
+    precompile_execution_file = joinpath(
+        @__DIR__,
+        "../../test/Numerics/DGmethods/Euler/isentropicvortex.jl",
```
Comment (charleskawczynski) [inline .dev/systemimage/climate_machine_image.jl:25]:
```suggestion
        "..","..","test","Numerics","DGmethods","Euler","isentropicvortex.jl",
```

File: .dev/systemimage/climate_machine_image.jl [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+#
+# Called with a single argument the system image will be placed in the path
+# specified by the argument (relative to the callers path)
+sysimage_path =
+    isempty(ARGS) ? joinpath(@__DIR__, "ClimateMachine.so") : ARGS[1]
+
+using Pkg
+Pkg.add("PackageCompiler")
+using PackageCompiler
+Pkg.activate(joinpath(@__DIR__, "../../"))
```
Comment (charleskawczynski) [inline .dev/systemimage/climate_machine_image.jl:15]:
```suggestion
Pkg.activate(joinpath(@__DIR__, "..",".."))
```

**Comment by jakebolewski** [comment]
Probably should add a sanity check test as well to make sure we catch instances when this breaks on PR's.

**Comment by jkozdon** [comment]
> IMO this should be in v0.1.

For users (as opposed to folks not hacking the CLIMA package itself), you could include a version that just depends on CLIMA itself, though there you would want to use as a precompile script something that uses the driver.

**Comment by jkozdon** [comment]
> Probably should add a sanity check test as well to make sure we catch instances when this breaks on PR's.

How are you thinking that this could be done?

**Comment by jakebolewski** [comment]
Maybe just a sanity check for now, invoke the command to run the script and make sure it doesn't segfault.

**Comment by jkozdon** [comment]
> Maybe just a sanity check for now, invoke the command / subprocess to run the script and make sure it doesn't segfault.

Not sure what you mean.

**Comment by jakebolewski** [comment]
@jkozdon I guess my question is if we want to ensure that this continues to work going forward (ex. include a test that precompilation is successful on CI), or if it's considered experimental for now.

**Comment by jkozdon** [comment]
> @jkozdon I guess my question is if we want to ensure that this continues to work going forward (ex. include a test that precompilation / creating a custom system image is successful on CI), or if it's considered experimental for now.

Something like https://github.com/CliMA/ClimateMachine.jl/pull/1056/commits/18b4bb5f57ced4b9638309717d23dcd2c35da711?

I don't think we want to actually build the full system image in the CI, but I could be wrong.

**Comment by jkozdon** [comment]
Possible the new MPI killed this.

Investigating more locally...

**Comment by jkozdon** [comment]
> Possible the new MPI killed this.
> 
> Investigating more locally...

nevermind, was a personal issue

bor r+

**Pull request merged**